File: python/src/kyccentral/_transport.py
"""HTTP plumbing shared by the sync and async clients.

Everything here is internal: the public surface is :mod:`kyccentral.client` and
:mod:`kyccentral.errors`. Kept in one module so that retry policy, error mapping
and header construction cannot drift between the two client flavours.
"""

from __future__ import annotations

import asyncio
import os
import platform
import random
import time
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any
from urllib.parse import quote

import httpx

from ._version import __version__
from .errors import APIConnectionError, APITimeoutError, status_error_from_response

DEFAULT_BASE_URL = "https://api.kyccentral.co.uk"
"""Production API. Override with ``base_url=`` or ``KYCCENTRAL_BASE_URL``."""

API_PREFIX = "/v1"
"""Every business endpoint is versioned. ``/health`` deliberately is not."""

DEFAULT_TIMEOUT = 30.0
DEFAULT_MAX_RETRIES = 2

_RETRYABLE_STATUSES = frozenset({408, 429, 500, 502, 503, 504})
_INITIAL_BACKOFF = 0.5
_MAX_BACKOFF = 8.0

API_KEY_ENV = "KYCCENTRAL_API_KEY"
BASE_URL_ENV = "KYCCENTRAL_BASE_URL"


def _default_user_agent() -> str:
    return (
        f"kyccentral-python/{__version__} "
        f"(httpx/{httpx.__version__}; python/{platform.python_version()})"
    )


def _clean_params(params: Mapping[str, Any] | None) -> dict[str, Any]:
    """Drop unset query parameters.

    ``None`` means "caller did not supply this", which must not be sent as the
    literal string ``None``. Empty sequences are dropped too: the API's
    repeatable parameters (``confirmed_media_url``, …) default to empty anyway,
    and sending nothing keeps URLs short.
    """
    if not params:
        return {}
    cleaned: dict[str, Any] = {}
    for key, value in params.items():
        if value is None:
            continue
        if isinstance(value, (list, tuple)):
            values = [v for v in value if v is not None]
            if not values:
                continue
            cleaned[key] = values
        else:
            cleaned[key] = value
    return cleaned


def _retry_delay(attempt: int, response: httpx.Response | None) -> float:
    """Seconds to wait before the next attempt.

    Honours ``Retry-After`` when the API sends one (it does on 429), otherwise
    backs off exponentially with full jitter so that a fleet of clients recovering
    from the same outage does not retry in lockstep.
    """
    if response is not None:
        raw = response.headers.get("Retry-After")
        if raw:
            try:
                return max(0.0, min(float(raw), _MAX_BACKOFF))
            except ValueError:
                pass
    ceiling = min(_INITIAL_BACKOFF * (2**attempt), _MAX_BACKOFF)
    return random.uniform(ceiling / 2, ceiling)


def _decode(response: httpx.Response) -> Any:
    """Decode a response body, tolerating empty and non-JSON payloads."""
    if response.status_code == 204 or not response.content:
        return None
    content_type = response.headers.get("content-type", "")
    if "json" in content_type:
        try:
            return response.json()
        except ValueError:
            return response.text
    return response.text


@dataclass(frozen=True)
class ClientConfig:
    """Resolved connection settings shared by a client and its resources."""

    api_key: str | None
    base_url: str
    timeout: float
    max_retries: int
    default_headers: Mapping[str, str]

    @classmethod
    def resolve(
        cls,
        api_key: str | None,
        base_url: str | None,
        timeout: float,
        max_retries: int,
        default_headers: Mapping[str, str] | None,
    ) -> ClientConfig:
        resolved_key = api_key if api_key is not None else os.environ.get(API_KEY_ENV) or None
        resolved_base = base_url or os.environ.get(BASE_URL_ENV) or DEFAULT_BASE_URL
        if max_retries < 0:
            raise ValueError("max_retries must be >= 0")
        if timeout <= 0:
            raise ValueError("timeout must be > 0")
        return cls(
            api_key=resolved_key,
            base_url=resolved_base.rstrip("/"),
            timeout=timeout,
            max_retries=max_retries,
            default_headers=dict(default_headers or {}),
        )

    def url_for(self, path: str, versioned: bool = True) -> str:
        prefix = API_PREFIX if versioned else ""
        return f"{self.base_url}{prefix}/{path.lstrip('/')}"

    def headers(self) -> dict[str, str]:
        headers: dict[str, str] = {
            "Accept": "application/json",
            "User-Agent": _default_user_agent(),
        }
        headers.update(self.default_headers)
        if self.api_key:
            headers["X-API-Key"] = self.api_key
        return headers


class _BaseTransport:
    """Shared retry bookkeeping. Subclasses supply the actual I/O."""

    def __init__(self, config: ClientConfig) -> None:
        self._config = config

    @property
    def config(self) -> ClientConfig:
        return self._config

    def _should_retry(self, attempt: int, status_code: int) -> bool:
        return attempt < self._config.max_retries and status_code in _RETRYABLE_STATUSES

    def _finish(
        self,
        response: httpx.Response,
        method: str,
        url: str,
    ) -> Any:
        body = _decode(response)
        if response.is_success:
            return body
        raise status_error_from_response(
            response.status_code,
            body,
            response.headers,
            method=method,
            url=url,
        )


class SyncTransport(_BaseTransport):
    """Blocking transport backed by :class:`httpx.Client`."""

    def __init__(self, config: ClientConfig, http_client: httpx.Client | None = None) -> None:
        super().__init__(config)
        self._owns_client = http_client is None
        self._http = http_client or httpx.Client(timeout=config.timeout, follow_redirects=True)

    def request(
        self,
        method: str,
        path: str,
        *,
        params: Mapping[str, Any] | None = None,
        json: Any = None,
        versioned: bool = True,
    ) -> Any:
        url = self._config.url_for(path, versioned)
        headers = self._config.headers()
        query = _clean_params(params)

        for attempt in range(self._config.max_retries + 1):
            try:
                response = self._http.request(method, url, params=query, json=json, headers=headers)
            except httpx.TimeoutException as exc:
                if attempt >= self._config.max_retries:
                    raise APITimeoutError(
                        f"{method} {url} timed out after {attempt + 1} attempt(s)."
                    ) from exc
                time.sleep(_retry_delay(attempt, None))
                continue
            except httpx.TransportError as exc:
                if attempt >= self._config.max_retries:
                    raise APIConnectionError(f"{method} {url} failed: {exc}") from exc
                time.sleep(_retry_delay(attempt, None))
                continue

            if self._should_retry(attempt, response.status_code):
                time.sleep(_retry_delay(attempt, response))
                continue
            return self._finish(response, method, url)

        raise APIConnectionError(f"{method} {url} exhausted all retries.")  # pragma: no cover

    def close(self) -> None:
        if self._owns_client:
            self._http.close()


class AsyncTransport(_BaseTransport):
    """Non-blocking transport backed by :class:`httpx.AsyncClient`."""

    def __init__(self, config: ClientConfig, http_client: httpx.AsyncClient | None = None) -> None:
        super().__init__(config)
        self._owns_client = http_client is None
        self._http = http_client or httpx.AsyncClient(timeout=config.timeout, follow_redirects=True)

    async def request(
        self,
        method: str,
        path: str,
        *,
        params: Mapping[str, Any] | None = None,
        json: Any = None,
        versioned: bool = True,
    ) -> Any:
        url = self._config.url_for(path, versioned)
        headers = self._config.headers()
        query = _clean_params(params)

        for attempt in range(self._config.max_retries + 1):
            try:
                response = await self._http.request(
                    method, url, params=query, json=json, headers=headers
                )
            except httpx.TimeoutException as exc:
                if attempt >= self._config.max_retries:
                    raise APITimeoutError(
                        f"{method} {url} timed out after {attempt + 1} attempt(s)."
                    ) from exc
                await asyncio.sleep(_retry_delay(attempt, None))
                continue
            except httpx.TransportError as exc:
                if attempt >= self._config.max_retries:
                    raise APIConnectionError(f"{method} {url} failed: {exc}") from exc
                await asyncio.sleep(_retry_delay(attempt, None))
                continue

            if self._should_retry(attempt, response.status_code):
                await asyncio.sleep(_retry_delay(attempt, response))
                continue
            return self._finish(response, method, url)

        raise APIConnectionError(f"{method} {url} exhausted all retries.")  # pragma: no cover

    async def aclose(self) -> None:
        if self._owns_client:
            await self._http.aclose()


class SyncResource:
    """Base class for blocking resource namespaces."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def _get(self, path: str, **kwargs: Any) -> Any:
        return self._transport.request("GET", path, **kwargs)

    def _post(self, path: str, **kwargs: Any) -> Any:
        return self._transport.request("POST", path, **kwargs)


class AsyncResource:
    """Base class for awaitable resource namespaces."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def _get(self, path: str, **kwargs: Any) -> Any:
        return await self._transport.request("GET", path, **kwargs)

    async def _post(self, path: str, **kwargs: Any) -> Any:
        return await self._transport.request("POST", path, **kwargs)


def _seg(value: str, name: str) -> str:
    """Validate and percent-encode a value being interpolated into a URL path.

    Empty segments would silently collapse the path onto a different endpoint,
    and officer ids are opaque upstream tokens, so both are checked and quoted
    rather than trusted.
    """
    if value is None or not str(value).strip():
        raise ValueError(f"{name} must be a non-empty string")
    return quote(str(value).strip(), safe="")


def _names_body(names: Sequence[str], *, limit: int | None = None) -> dict[str, Any]:
    """Build the ``{"names": [...]}`` body used by the batch screening endpoints."""
    cleaned = [str(n).strip() for n in names if str(n).strip()]
    if not cleaned:
        raise ValueError("names must contain at least one non-empty string")
    if limit is not None and len(cleaned) > limit:
        raise ValueError(f"names accepts at most {limit} entries, got {len(cleaned)}")
    return {"names": cleaned}

File: elixir/lib/kyccentral/transport.ex
defmodule KYCCentral.Transport do
  @moduledoc """
  HTTP plumbing: URL building, retries, decoding and error mapping.

  Internal. The public surface is `KYCCentral` and the resource modules.
  """

  alias KYCCentral.Error

  @retryable_statuses [408, 429, 500, 502, 503, 504]
  @initial_backoff_ms 500
  @max_backoff_ms 8_000

  @doc """
  Issue a request, retrying transient failures.

  Returns the decoded body on success, or `{:error, %KYCCentral.Error{}}`.
  """
  @spec request(KYCCentral.t(), :get | :post, String.t(), keyword()) ::
          {:ok, term()} | {:error, Error.t()}
  def request(%KYCCentral{} = client, method, path, opts \\ []) do
    url = build_url(client, path, opts[:params], Keyword.get(opts, :versioned, true))
    attempt(client, method, url, opts[:body], 0)
  end

  defp attempt(client, method, url, body, retries) do
    verb = method |> to_string() |> String.upcase()

    case perform(client, method, url, body) do
      {:ok, %{status: status} = response} when status in @retryable_statuses ->
        if retries < client.max_retries do
          response |> retry_delay(retries) |> sleep()
          attempt(client, method, url, body, retries + 1)
        else
          finish(response, verb, url)
        end

      {:ok, response} ->
        finish(response, verb, url)

      {:error, reason} ->
        if retries < client.max_retries do
          sleep(retry_delay(nil, retries))
          attempt(client, method, url, body, retries + 1)
        else
          {:error, transport_error(reason, verb, url)}
        end
    end
  end

  defp perform(client, method, url, body) do
    request = %{
      method: method,
      url: url,
      headers: headers(client, body != nil),
      body: if(body, do: Jason.encode!(body)),
      receive_timeout: client.receive_timeout
    }

    client.http.(request)
  end

  @doc """
  The default HTTP function, backed by OTP's own `:httpc`.

  Using `:httpc` is what lets this library ship with Jason as its only runtime
  dependency — adding it to a project pulls in no HTTP stack. Pass `:http` to
  `KYCCentral.new/1` to use Req, Finch, Tesla or anything else instead.

  Retries are handled by this module rather than the transport, so the policy
  matches the Python and JavaScript clients. Bodies are returned undecoded and
  decoded centrally, so an injected function and the default behave identically.
  """
  @spec httpc_http(map()) :: {:ok, map()} | {:error, term()}
  def httpc_http(request) do
    {:ok, _started} = Application.ensure_all_started(:inets)
    {:ok, _started} = Application.ensure_all_started(:ssl)

    url = String.to_charlist(request.url)

    http_options = [
      timeout: request.receive_timeout,
      connect_timeout: request.receive_timeout,
      ssl: ssl_options()
    ]

    options = [body_format: :binary]

    result =
      case request.method do
        :get ->
          :httpc.request(:get, {url, charlist_headers(request.headers)}, http_options, options)

        :post ->
          # httpc takes the content type as its own argument and rejects a
          # duplicate in the header list, so it is filtered out here.
          headers = Enum.reject(request.headers, &(elem(&1, 0) == "content-type"))

          :httpc.request(
            :post,
            {url, charlist_headers(headers), ~c"application/json", request.body || ""},
            http_options,
            options
          )
      end

    case result do
      {:ok, {{_version, status, _reason}, headers, body}} ->
        {:ok, %{status: status, headers: decode_headers(headers), body: body}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  # `:httpc` does not verify certificates unless told to, and a client that
  # carries an API key must never talk to an unverified peer. `cacerts_get/0`
  # uses the OS trust store (OTP 25+).
  defp ssl_options do
    [
      verify: :verify_peer,
      cacerts: :public_key.cacerts_get(),
      depth: 3,
      customize_hostname_check: [
        match_fun: :public_key.pkix_verify_hostname_match_fun(:https)
      ],
      versions: [:"tlsv1.2", :"tlsv1.3"]
    ]
  end

  defp charlist_headers(headers) do
    Enum.map(headers, fn {key, value} ->
      {String.to_charlist(to_string(key)), String.to_charlist(to_string(value))}
    end)
  end

  defp decode_headers(headers) do
    Map.new(headers, fn {key, value} -> {to_string(key), to_string(value)} end)
  end

  defp headers(client, has_body?) do
    base = [{"accept", "application/json"}, {"user-agent", user_agent()}]
    base = if has_body?, do: [{"content-type", "application/json"} | base], else: base
    base = if client.api_key, do: [{"x-api-key", client.api_key} | base], else: base
    base ++ Enum.map(client.headers, fn {key, value} -> {to_string(key), to_string(value)} end)
  end

  defp user_agent do
    version = Application.spec(:kyccentral, :vsn) || ~c"dev"
    "kyccentral-elixir/#{version} (elixir/#{System.version()}; otp/#{System.otp_release()})"
  end

  defp finish(%{status: status, headers: headers, body: body}, verb, url) do
    decoded = decode(body, headers)

    if status in 200..299 do
      {:ok, decoded}
    else
      {:error, Error.from_response(status, decoded, normalise_headers(headers), verb, url)}
    end
  end

  # Tolerates empty and non-JSON payloads, and accepts an already-decoded body so
  # a test stub can hand back a map directly.
  defp decode(body, _headers) when is_map(body) or is_list(body), do: body
  defp decode(nil, _headers), do: nil
  defp decode("", _headers), do: nil

  defp decode(body, headers) when is_binary(body) do
    content_type = Error.header(headers, "content-type") || ""

    if String.contains?(content_type, "json") do
      case Jason.decode(body) do
        {:ok, decoded} -> decoded
        {:error, _} -> body
      end
    else
      body
    end
  end

  defp decode(body, _headers), do: body

  defp normalise_headers(headers) when is_map(headers), do: headers
  defp normalise_headers(headers) when is_list(headers), do: Map.new(headers)
  defp normalise_headers(_), do: %{}

  defp transport_error(reason, verb, url) do
    if timeout?(reason) do
      %Error{
        kind: :timeout,
        message: "#{verb} #{url} timed out.",
        reason: reason
      }
    else
      %Error{
        kind: :connection,
        message: "#{verb} #{url} failed: #{describe(reason)}",
        reason: reason
      }
    end
  end

  defp timeout?(:timeout), do: true
  defp timeout?({:failed_connect, details}) when is_list(details), do: nested_timeout?(details)
  defp timeout?(%{reason: :timeout}), do: true
  defp timeout?(_), do: false

  # `:httpc` reports a connect timeout as
  # `{:failed_connect, [{:to_address, _}, {:inet, [:inet], :timeout}]}`.
  defp nested_timeout?(details) do
    Enum.any?(details, fn
      {_transport, _options, :timeout} -> true
      {_key, :timeout} -> true
      _ -> false
    end)
  end

  defp describe(%{__exception__: true} = exception), do: Exception.message(exception)
  defp describe(reason), do: inspect(reason)

  # Honours `Retry-After` when the API sends one (it does on 429), otherwise backs
  # off exponentially with full jitter, so a fleet recovering from one outage does
  # not retry in lockstep.
  defp retry_delay(%{headers: headers}, attempt) do
    case Error.retry_after(headers) do
      nil -> jittered_backoff(attempt)
      seconds -> seconds |> Kernel.*(1000) |> round() |> min(@max_backoff_ms) |> max(0)
    end
  end

  defp retry_delay(nil, attempt), do: jittered_backoff(attempt)

  defp jittered_backoff(attempt) do
    ceiling = min(@initial_backoff_ms * 2 ** attempt, @max_backoff_ms)
    round(ceiling / 2 + :rand.uniform() * (ceiling / 2))
  end

  defp sleep(0), do: :ok
  defp sleep(milliseconds), do: Process.sleep(milliseconds)

  @doc false
  @spec build_url(KYCCentral.t(), String.t(), keyword() | nil, boolean()) :: String.t()
  def build_url(client, path, params, versioned?) do
    prefix = if versioned?, do: "/v1", else: ""
    path = if String.starts_with?(path, "/"), do: path, else: "/" <> path
    client.base_url <> prefix <> path <> query_string(params)
  end

  @doc """
  Encode query parameters, dropping unset ones and repeating list values.

  Built by hand rather than through `URI.encode_query/1` because several
  endpoints take repeatable parameters (`confirmed_media_url`, `company_status`),
  which need `?k=a&k=b` rather than one comma-joined value.
  """
  @spec query_string(keyword() | nil) :: String.t()
  def query_string(nil), do: ""
  def query_string([]), do: ""

  def query_string(params) do
    case Enum.flat_map(params, &encode_param/1) do
      [] -> ""
      pairs -> "?" <> Enum.join(pairs, "&")
    end
  end

  # `nil` means "caller did not supply this" and must never be sent as the
  # literal string "nil".
  defp encode_param({_key, nil}), do: []
  defp encode_param({_key, []}), do: []

  defp encode_param({key, values}) when is_list(values) do
    values |> Enum.reject(&is_nil/1) |> Enum.map(&pair(key, &1))
  end

  defp encode_param({key, value}), do: [pair(key, value)]

  defp pair(key, value) do
    "#{URI.encode_www_form(to_string(key))}=#{URI.encode_www_form(to_string(value))}"
  end

  @doc """
  Validate and percent-encode a value being interpolated into a URL path.

  Empty segments would silently collapse the path onto a different endpoint, and
  officer ids are opaque upstream tokens, so both are checked and encoded.
  """
  @spec segment(term(), String.t()) :: {:ok, String.t()} | {:error, Error.t()}
  def segment(value, name) when is_binary(value) do
    case String.trim(value) do
      "" -> {:error, Error.invalid_argument("#{name} must be a non-empty string")}
      trimmed -> {:ok, URI.encode_www_form(trimmed)}
    end
  end

  def segment(_value, name) do
    {:error, Error.invalid_argument("#{name} must be a non-empty string")}
  end

  @doc """
  Build the `%{"names" => [...]}` body used by the batch screening endpoints.
  """
  @spec names_body([String.t()], pos_integer() | nil) :: {:ok, map()} | {:error, Error.t()}
  def names_body(names, limit \\ nil) when is_list(names) do
    cleaned =
      names
      |> Enum.map(&(&1 |> to_string() |> String.trim()))
      |> Enum.reject(&(&1 == ""))

    cond do
      cleaned == [] ->
        {:error, Error.invalid_argument("names must contain at least one non-empty string")}

      limit && length(cleaned) > limit ->
        {:error,
         Error.invalid_argument("names accepts at most #{limit} entries, got #{length(cleaned)}")}

      true ->
        {:ok, %{"names" => cleaned}}
    end
  end
end
